### ISSUE TITLE
LeRobot DP policy trained with official config produces static robot during eval videos

### ISSUE THREAD

**Issue by 1024AILab** [lehome-official/lehome-challenge#69]
## Summary

I followed the official LeHome Challenge training and evaluation workflow to train a LeRobot Diffusion Policy (DP). Training completes successfully, and `scripts.eval` can launch the Isaac/LeHome environment and save evaluation videos.

However, in the saved MP4 files, the bimanual SO101 robot arms appear to stay still or almost still. The episode runs until `max_steps=600`, videos are saved under the `failure` folder, and the success checker repeatedly reports failure.

I would like to ask whether this is caused by an expected configuration requirement, a mismatch between the official DP training config and the evaluation action interface, or a bug in the LeRobot policy evaluation pipeline.

## Repository

- Repository: https://github.com/lehome-official/lehome-challenge
- Local path: `/root/autodl-tmp/lehome-challenge`
- Commit: `2100362`

## Platform And Environment

- Platform: AutoDL
- OS: Linux container
- Mode: headless Isaac Sim / IsaacLab
- GPU: NVIDIA GeForce RTX 5090
- CUDA: 12.8
- Python: 3.11
- PyTorch: `2.7.0+cu128`
- Torch CUDA arch list includes `sm_120`

I am using the existing environment from:

```bash
/root/autodl-tmp/lehome/.venv
```

Activated with:

```bash
source /root/autodl-tmp/lehome/.venv/bin/activate
```

## Dataset

I trained and evaluated on the merged four-category dataset:

```bash
/root/autodl-tmp/lehome/Datasets/example/four_types_merged
```

## Training Configuration

I trained a LeRobot Diffusion Policy with a config based on the official DP training config.

Important fields:

```yaml
dataset:
  repo_id: lehome_fold_garment
  root: /root/autodl-tmp/lehome/Datasets/example/four_types_merged

policy:
  type: diffusion
  device: cuda
  push_to_hub: false

  crop_shape: null
  crop_is_random: false

  input_features:
    observation.state:
      type: STATE
      shape: [12]
    observation.images.top_rgb:
      type: VISUAL
      shape: [3, 480, 640]
    observation.images.left_rgb:
      type: VISUAL
      shape: [3, 480, 640]
    observation.images.right_rgb:
      type: VISUAL
      shape: [3, 480, 640]

  output_features:
    action:
      type: ACTION
      shape: [12]

output_dir: outputs/train/dp_so101_test
batch_size: 16
steps: 30000
save_freq: 5000
log_freq: 1000

wandb:
  enable: false
```

The training checkpoint path is:

```bash
/root/autodl-tmp/lehome/outputs/train/dp_so101_test/checkpoints/last/pretrained_model
```

The `last` checkpoint points to step `030000`.

## Evaluation Command

I ran evaluation from the `lehome-challenge` repository:

```bash
cd /root/autodl-tmp/lehome-challenge
source /root/autodl-tmp/lehome/.venv/bin/activate

python -m scripts.eval \
  --task LeHome-BiSO101-Direct-Garment-v2 \
  --policy_type lerobot \
  --policy_path /root/autodl-tmp/lehome/outputs/train/dp_so101_test/checkpoints/last/pretrained_model \
  --dataset_root /root/autodl-tmp/lehome/Datasets/example/four_types_merged \
  --task_description "fold the garment on the table" \
  --num_episodes 5 \
  --enable_cameras \
  --device cuda \
  --headless \
  --save_video \
  --video_dir outputs/eval_videos/dp_so101_test
```

The evaluation starts successfully and saves MP4 videos. There is no crash during policy loading or video saving.

## Observed Behavior

- The evaluation runs until `Length=600`.
- The saved MP4 videos show that the robot arms remain still or almost still.
- The videos are saved under:

```bash
outputs/eval_videos/dp_so101_test/failure/
```

Example saved videos:

```bash
outputs/eval_videos/dp_so101_test/failure/episode2_observation_images_top_rgb.mp4
outputs/eval_videos/dp_so101_test/failure/episode2_observation_images_left_rgb.mp4
outputs/eval_videos/dp_so101_test/failure/episode2_observation_images_right_rgb.mp4
```

- The success checker repeatedly reports failure.
- The garment distances change only slightly over time.
- The robot does not appear to execute the learned folding motion.

Example log:

```text
2026-05-04 17:57:41 - lehome.tasks.bedroom.garment_bi_v2 - INFO - [Success Check] Garment type: top-long-sleeve, Thresholds: [7.2, 8.55, 7.2, 9.9, 9.0]
2026-05-04 17:57:41 - lehome.tasks.bedroom.garment_bi_v2 - INFO -   dist(p[0], p[4]) = 24.33 <= 7.2 -> FAIL
2026-05-04 17:57:41 - lehome.tasks.bedroom.garment_bi_v2 - INFO -   dist(p[2], p[3]) = 45.79 <= 8.55 -> FAIL
2026-05-04 17:57:41 - lehome.tasks.bedroom.garment_bi_v2 - INFO -   dist(p[1], p[5]) = 25.58 <= 7.2 -> FAIL
2026-05-04 17:57:41 - lehome.tasks.bedroom.garment_bi_v2 - INFO -   dist(p[0], p[1]) = 19.62 >= 9.9 -> PASS
2026-05-04 17:57:41 - lehome.tasks.bedroom.garment_bi_v2 - INFO -   dist(p[4], p[5]) = 14.10 >= 9.0 -> PASS
2026-05-04 17:57:41 - lehome.tasks.bedroom.garment_bi_v2 - INFO - [Success Check] Final result: Failed

2026-05-04 17:58:47 - scripts.utils.eval_utils - INFO - Saved video: outputs/eval_videos/dp_so101_test/failure/episode2_observation_images_top_rgb.mp4
2026-05-04 17:58:48 - scripts.utils.eval_utils - INFO - Saved video: outputs/eval_videos/dp_so101_test/failure/episode2_observation_images_left_rgb.mp4
2026-05-04 17:58:49 - scripts.utils.eval_utils - INFO - Saved video: outputs/eval_videos/dp_so101_test/failure/episode2_observation_images_right_rgb.mp4
2026-05-04 17:58:49 - scripts.utils.evaluation - INFO - Episode 3/5: Return=108.03, Length=600, Success=False
```

## Expected Behavior

After training a DP policy with the official config and evaluating it with `scripts.eval --policy_type lerobot`, the policy should produce actions that visibly move the bimanual SO101 robot arms in the saved evaluation videos.

If additional settings are required for DP evaluation, it would be helpful to document them in the README or training/evaluation docs.

## Things I Already Checked

- The environment can launch successfully.
- The task name `LeHome-BiSO101-Direct-Garment-v2` is used instead of an invalid task name such as `fold_garment`.
- The policy checkpoint path exists.
- Videos are saved correctly.
- The evaluation loop runs to completion for each episode.
- The RTX 5090 CUDA architecture issue was fixed by using a PyTorch build that supports `sm_120`.

## Questions

1. Is `--device cuda` supported for `scripts.eval`? The README mentions that simulation currently only supports CPU devices, but policy inference may still use CUDA.
2. Should `scripts.eval` always be run with `--device cpu` for the simulation, even when evaluating a CUDA-trained policy?
3. Is there any required action scaling, action clipping, normalization, or post-processing step for LeRobot DP policies that may be missing from the current eval pipeline?
4. Is the official DP training config expected to work with `scripts.eval --policy_type lerobot` out of the box?
5. Is `LeHome-BiSO101-Direct-Garment-v2` the correct eval task for DP policies trained on the merged four-category dataset?
6. Are there any known issues where LeRobot DP checkpoints load successfully but produce near-zero or ineffective robot actions during eval?

## Attachments

I can provide the saved MP4 files showing the static robot behavior:

```bash
outputs/eval_videos/dp_so101_test/failure/episode2_observation_images_top_rgb.mp4
outputs/eval_videos/dp_so101_test/failure/episode2_observation_images_left_rgb.mp4
outputs/eval_videos/dp_so101_test/failure/episode2_observation_images_right_rgb.mp4
```

I can also provide the full training config, full eval logs, and checkpoint config if needed.

## Additional Note

If this is caused by using `--device cuda` for simulation, it would be very helpful to clarify the recommended command. For example, should the official eval command be:

```bash
python -m scripts.eval \
  --task LeHome-BiSO101-Direct-Garment-v2 \
  --policy_type lerobot \
  --policy_path /path/to/pretrained_model \
  --dataset_root /path/to/dataset \
  --task_description "fold the garment on the table" \
  --num_episodes 5 \
  --enable_cameras \
  --device cpu \
  --headless \
  --save_video
```

instead of using `--device cuda`?

https://github.com/user-attachments/assets/1039e4e2-f370-4b20-a9e2-882e00a07c9d
https://github.com/user-attachments/assets/c5bb37d1-814b-40ae-975f-20ab8c299594
https://github.com/user-attachments/assets/8563d9f2-e45f-4624-b44e-8959ad70656c

**Comment by davechendatascience** [comment]
Seeing the same failure on different hardware, with some measurements that may
narrow it down.

## Environment

| | |
|---|---|
| GPU | NVIDIA GB10 (DGX Spark), sm_121, **aarch64** |
| Isaac Sim | 5.1.0.0 |
| IsaacLab | `lehome-official/IsaacLab` @ `69f6fa54` |
| torch | 2.7.0+cu128, Python 3.11 |
| dataset | `Datasets/example/top_long_merged`, garment `Top_Long_Seen_0` |

Same symptom as the original report: arms move, garment does not, episodes run to
`max_steps`, videos land in `failure/`, checker reports failure.

## The measurement that seems most useful

Replaying the **recorded actions** from the dataset (feeding
`episode 0`'s recorded `action` values straight into `env.step`) *does* fold the
garment. Trained policies do not. Same environment, same garment, same episode.

Check-point displacement, measured after the garment has settled (the reset drop
from z=0.73 to z=0.53 must be excluded or it swamps everything):

| driver | p0 | p1 | p2 | p3 | p4 | p5 |
|---|---|---|---|---|---|---|
| **replayed recorded actions** | 1.1 | 0.6 | 0.7 | 0.2 | **29.8** | **26.6** |
| SmolVLA finetuned from `smolvla_base`, 15k steps | 0.7 | 0.5 | 0.5 | 0.3 | 1.1 | 0.4 |
| `Papercold/lehome-smolvla-top-long` (30k steps, community) | 0.7 | 0.6 | 0.4 | 4.0 | 1.1 | 0.5 |
| `IliaLarchenko/lehome_sim` (1st place, pi0.5, reported 74.5% on long tops) | 0.6 | 0.6 | 0.5 | 6.3 | 1.1 | 0.5 |

Three independently trained policies, two architectures, one of them the released
winning checkpoint — all produce essentially no manipulation, while open-loop
replay works. That is what makes me think this is environmental rather than a
training problem.

Segmented garment footprint over an episode tells the same story
(Otsu threshold, robot masked out):

```
                  step 0%   36%    72%   100%
recorded dataset    20.5%  15.3%  11.3%   6.3%     <- folds
our sim, replayed   37.1%  38.5%  38.1%  38.5%     <- flat
```

## The mechanism I can see

Logging the distance from each gripper body to the nearest of the 14,544 cloth
particles, every step, while replaying recorded actions (965 ticks):

```
                 min     p05   median   steps within 0.09 m
left  gripper   0.065   0.068   0.110      156/965
right gripper   0.101   0.103   0.146        0/965
```

The gripper *body* origin sits ~9 cm from the actual contact point, so the left
hand reaches the fabric and **the right hand never does — 0 of 965 steps**, ~1 cm
short at its closest. A bimanual fold with one hand always off the cloth would
explain both the replay variability and why a closed-loop policy gets no useful
feedback.

Is the recorded dataset expected to reproduce the fold in the current sim? If the
recorded joint trajectories are supposed to land both grippers on the garment and
here one of them consistently misses, that would point at a robot/base/URDF
offset rather than anything policy-side.

## Framing question

Our top camera renders the garment noticeably larger than the recorded frames do:
37.1% of frame area against the dataset's 20.5% (measured over three episodes:
20.5 / 20.5 / 24.5), with the garment clipped at the frame edges. Camera geometry
reads as configured — world position (-0.015, 0.190, 1.060), garment centroid
(-0.008, 0.056, 0.530), distance 0.547 m, HFOV 67.2 deg — so the config is being
applied; the recorded data just implies a camera roughly 0.27 m further back.
Is a framing difference between the released dataset and the current eval config
expected?

## Ruled out here

In case it saves anyone time — none of these changed the outcome:

* `--rendering_mode` performance / balanced / quality
* `SimulationCfg.use_fabric` True and False
* **`TiledCamera` -> `Camera`** (the fix from #59). Byte-identical output for us,
  and we never saw the `unsupported toolchain` / warp-compile error, so #59 does
  not appear to apply to GB10 even though it is also Blackwell
* camera registration (cameras are in `scene.sensors` and do update)
* geometry writeback — `get_current_mesh_points()` and
  `_cloth_prim_view.get_world_positions()` agree exactly, and both move
* action units — the policy returns absolute joint targets, matching the dataset
* running the winner's own fork of this repo for env + eval code

The render is faithful to the physics: the rendered garment and its 3-D bbox
agree with each other. The cloth genuinely is not being manipulated.

## Two small bugs found on the way

1. `success_checker_garment_fold` is decorated `@step_interval(interval=50)`,
   which returns a literal `False` on 49 of 50 calls where callers expect the
   result dict, and its `call_count` lives in the decorator closure so it never
   resets between episodes or garment switches. The evaluated step therefore
   drifts arbitrarily through a run and a transient success can be missed. Our
   episodes logged 13 checks over 600 steps before we bypassed it, 315 after.

2. `/isaaclab/cameras_enabled` is commented out in `tiled_camera.py` in the
   IsaacLab fork but still enforced in `camera.py`. So `--enable_cameras` never
   actually reaches the setting, and swapping to `Camera` per #59 fails with
   "A camera was spawned without the --enable_cameras flag" until that guard is
   disabled too.

Happy to share the probe scripts or run further diagnostics if useful.
